File: OsmAnd-java/src/main/java/net/osmand/binary/ObfConstants.java
package net.osmand.binary;

import static net.osmand.data.MapObject.AMENITY_ID_RIGHT_SHIFT;

import java.util.Locale;
import java.util.Map;

import net.osmand.NativeLibrary.RenderedObject;
import net.osmand.data.Amenity;
import net.osmand.data.MapObject;
import net.osmand.osm.edit.Entity;
import net.osmand.osm.edit.Entity.EntityType;
import net.osmand.osm.edit.Relation;
import net.osmand.util.Algorithms;

public class ObfConstants {

	public static final int SHIFT_ID = 6;
	
	public static final int SHIFT_MULTIPOLYGON_IDS = 43;
	public static final int SHIFT_NON_SPLIT_EXISTING_IDS = 41;

	public static final int SHIFT_PROPAGATED_NODE_IDS = 50;
	public static final int SHIFT_PROPAGATED_NODES_BITS = 11;
	public static final long MAX_ID_PROPAGATED_NODES = (1L << SHIFT_PROPAGATED_NODES_BITS) - 1;//2047

	public static final long RELATION_BIT = 1L << (ObfConstants.SHIFT_MULTIPOLYGON_IDS - 1); // 1L << 42
	public static final long PROPAGATE_NODE_BIT = 1L << (ObfConstants.SHIFT_PROPAGATED_NODE_IDS  - 1); // 1L << 49
	public static final long SPLIT_BIT = 1L << (ObfConstants.SHIFT_NON_SPLIT_EXISTING_IDS - 1); // 1L << 40

	public static final int DUPLICATE_SPLIT = 5;

	public static String getOsmUrlForId(MapObject mapObject) {
		EntityType type = getOsmEntityType(mapObject);
		if (type != null) {
			long osmId = getOsmObjectId(mapObject);
			return "https://www.openstreetmap.org/" + type.name().toLowerCase(Locale.US) + "/" + osmId;
		}
		return "";
	}

	public static long getOsmIdFromPrefixedRouteId(String routeId) {
		long osmId = 0;
		if (routeId.startsWith(Amenity.ROUTE_ID_OSM_PREFIX)) {
			osmId = Algorithms.parseLongSilently(routeId.replace(Amenity.ROUTE_ID_OSM_PREFIX, ""), 0); // ^O
		} else if (routeId.startsWith(Amenity.ROUTE_ID_OSM_PREFIX_LEGACY)) {
			osmId = Algorithms.parseLongSilently(routeId.replace(Amenity.ROUTE_ID_OSM_PREFIX_LEGACY, ""), 0); // ^OSM
		}
		return osmId;
	}
	
	
	public static long createMapObjectIdFromOsmAndEntity(Entity e) {
		// Way / Node with shifts , Relation without
		boolean isPropagated = e.getId() > PROPAGATE_NODE_BIT;
		if (isPropagated) {
			return e.getId();
		}
		long cleanId = e instanceof Relation ? e.getId() : (e.getId() >> SHIFT_ID);
		return createMapObjectIdFromCleanOsmId(cleanId, EntityType.valueOf(e));
	}

	public static long createMapObjectIdFromCleanOsmId(long osmId, EntityType type) {
		if (type == null) {
			return osmId;
		}
		return switch (type) {
			case NODE -> osmId << AMENITY_ID_RIGHT_SHIFT;
			case WAY -> (osmId << AMENITY_ID_RIGHT_SHIFT) + 1;
			case RELATION -> RELATION_BIT + ((osmId << SHIFT_ID) << DUPLICATE_SPLIT);
			default -> osmId;
		};
	}

	public static long getOsmObjectId(MapObject object) {
		long originalId = -1;
		Long id = object.getId();
		if (id != null) {
			if (object instanceof RenderedObject) {
				id = shiftRenderedObjectId(id);
			}
			originalId = getOsmIdFromMapObjectId(id);
		}
		return originalId;
	}

	private static long shiftRenderedObjectId(long id) {
		return isIdFromPropagatedNode(id >> 1) ? id >> 1 : id >> SHIFT_ID;
	}

	// Use getOsmObjectId(MapObject) for RenderedObject
	public static long getOsmIdFromMapObjectId(long id) {
		long originalId;
		if (isIdFromPropagatedNode(id)) {
			long shifted = id & ~PROPAGATE_NODE_BIT;
			originalId = shifted >> ObfConstants.SHIFT_PROPAGATED_NODES_BITS;
		} else {
			if (isShiftedID(id)) {
				originalId = getOsmId(id);
			} else {
				originalId = id >> AMENITY_ID_RIGHT_SHIFT;
			}
		}
		return originalId; 
	}
	
	public static long getOsmIdFromBinaryMapObjectId(long id) {
		// BinaryMapDataObject object
		return getOsmId(id >> 1); 
	}

	public static long getOsmObjectId(RouteDataObject object) {
		if (object == null) {
			return 0;
		}
		return getOsmId(object.getId());
	}
	
	public static long getOsmObjectId(BinaryMapDataObject object) {
		if (object == null) {
			return 0;
		}
		return getOsmId(object.getId() >> 1);
	}

	public static EntityType getOsmEntityType(MapObject object) {
		if (isOsmUrlAvailable(object)) {
			Long id = object.getId();
			long originalId = id >> 1;
			if (object instanceof RenderedObject && isIdFromPropagatedNode(originalId)) {
				return EntityType.WAY;
			}
			if (isIdFromPropagatedNode(id)) {
				return EntityType.WAY;
			}
			long relationShift = 1L << 41;
			if (originalId > relationShift) {
				return EntityType.RELATION;
			} else {
				return id % 2 == MapObject.WAY_MODULO_REMAINDER ? EntityType.WAY : EntityType.NODE;
			}
		}
		return null;
	}

	public static String getPrintTags(RenderedObject renderedObject) {
		StringBuilder s = new StringBuilder();
		for (Map.Entry<String, String> entry : renderedObject.getTags().entrySet()) {
			String key = entry.getKey();
			String value = entry.getValue();
			boolean keyEmpty = Algorithms.isEmpty(key);
			boolean valueEmpty = Algorithms.isEmpty(value);
			boolean bothPresent = !keyEmpty && !valueEmpty;
			boolean anyPresent = !keyEmpty || !valueEmpty;
			if (!keyEmpty) {
				s.append(key);
			}
			if (bothPresent) {
				s.append(":");
			}
			if (!valueEmpty) {
				s.append(value);
			}
			if (anyPresent) {
				s.append(" ");
			}
		}
		return s.toString().trim();
	}

	public static boolean isOsmUrlAvailable(MapObject object) {
		Long id = object.getId();
		return id != null && id > 0;
	}

	public static EntityType getOsmEntityType(BinaryMapDataObject object) {
		if(isIdFromRelation(object.getId())) {
			return EntityType.RELATION;
		}
		if(object.getId() % 2 == 0) {
			return EntityType.NODE;
		}
		return EntityType.WAY;
	}
	
	private static long getOsmId(long id) {
		//According methods assignIdForMultipolygon and genId in IndexPoiCreator
		long clearBits = RELATION_BIT | SPLIT_BIT;
		id = isShiftedID(id) ? (id & ~clearBits) >> DUPLICATE_SPLIT : id;
		return id >> SHIFT_ID;
	}

	private static boolean isShiftedID(long id) {
		return isIdFromRelation(id) || isIdFromSplit(id);
	}

	private static boolean isIdFromRelation(long id) {
		return id > 0 && (id & RELATION_BIT) == RELATION_BIT;
	}

	private static boolean isIdFromPropagatedNode(long id) {
		return id > 0 && (id & PROPAGATE_NODE_BIT) == PROPAGATE_NODE_BIT;
	}

	private static boolean isIdFromSplit(long id) {
		return id > 0 && (id & SPLIT_BIT) == SPLIT_BIT;
	}

	public static boolean isTagIndexedForSearchAsName(String tag) {
		if (tag != null) {
			return tag.contains("name") || tag.contains("brand");
		}
		return false;
	}
	
	public static boolean isTagIndexedForSearchAsId(String tag) {
		if (tag != null) {
			return tag.equals(Amenity.WIKIDATA) || tag.equals(Amenity.ROUTE_ID);
		}
		return false;
	}

}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/binary/ObfConstants.kt
package net.osmand.shared.binary

import net.osmand.shared.data.Amenity
import net.osmand.shared.data.MapObject
import net.osmand.shared.routing.RouteDataObject
import net.osmand.shared.util.KAlgorithms

/**
 * How an osm id is packed into an obf id, and which tags the indexer treats as names or as ids.
 *
 * A copy of the part of `ObfConstants` in OsmAnd-java that a device needs; the rest of it - the osm
 * url helpers, `EntityType` and everything taking a `RenderedObject` or an `Entity` - belongs to
 * the editing and indexing code and stays there.
 *
 * The packing is what the map creator does on the way in: a node id is shifted left by one, a way
 * id by one with the low bit set, and a relation id is shifted further and marked with
 * [RELATION_BIT]. Reading an object back gives the packed id, so anything that wants the osm id -
 * an url, a route id, a comparison against osm data - has to undo it here.
 */
object ObfConstants {

	const val SHIFT_ID: Int = 6

	const val SHIFT_MULTIPOLYGON_IDS: Int = 43
	const val SHIFT_NON_SPLIT_EXISTING_IDS: Int = 41

	const val SHIFT_PROPAGATED_NODE_IDS: Int = 50
	const val SHIFT_PROPAGATED_NODES_BITS: Int = 11
	const val MAX_ID_PROPAGATED_NODES: Long = (1L shl SHIFT_PROPAGATED_NODES_BITS) - 1 // 2047

	const val RELATION_BIT: Long = 1L shl (SHIFT_MULTIPOLYGON_IDS - 1) // 1L << 42
	const val PROPAGATE_NODE_BIT: Long = 1L shl (SHIFT_PROPAGATED_NODE_IDS - 1) // 1L << 49
	const val SPLIT_BIT: Long = 1L shl (SHIFT_NON_SPLIT_EXISTING_IDS - 1) // 1L << 40

	const val DUPLICATE_SPLIT: Int = 5

	/** The osm id behind a route id of the form `O123456`, or its legacy form `OSM123456`. */
	fun getOsmIdFromPrefixedRouteId(routeId: String): Long {
		var osmId = 0L
		if (routeId.startsWith(Amenity.ROUTE_ID_OSM_PREFIX)) {
			osmId = KAlgorithms.parseLongSilently(routeId.replace(Amenity.ROUTE_ID_OSM_PREFIX, ""), 0) // ^O
		} else if (routeId.startsWith(Amenity.ROUTE_ID_OSM_PREFIX_LEGACY)) {
			osmId = KAlgorithms.parseLongSilently(routeId.replace(Amenity.ROUTE_ID_OSM_PREFIX_LEGACY, ""), 0) // ^OSM
		}
		return osmId
	}

	fun getOsmObjectId(obj: MapObject): Long {
		var originalId = -1L
		val id = obj.getId()
		if (id != null) {
			originalId = getOsmIdFromMapObjectId(id)
		}
		return originalId
	}

	// Doesn't work correctly for some TransportStop (stop.getId() = 16055353830 results in osmId 8027676915 which does not exist)
	// https://www.openstreetmap.org/node/988560310
	fun getOsmIdFromMapObjectId(id: Long): Long {
		val originalId: Long
		if (isIdFromPropagatedNode(id)) {
			val shifted = id and PROPAGATE_NODE_BIT.inv()
			originalId = shifted shr SHIFT_PROPAGATED_NODES_BITS
		} else {
			originalId = if (isShiftedID(id)) {
				getOsmId(id)
			} else {
				id shr MapObject.AMENITY_ID_RIGHT_SHIFT
			}
		}
		return originalId
	}

	fun getOsmIdFromBinaryMapObjectId(id: Long): Long {
		// BinaryMapDataObject object
		return getOsmId(id shr 1)
	}

	fun getOsmObjectId(obj: RouteDataObject?): Long {
		if (obj == null) {
			return 0
		}
		return getOsmId(obj.getId())
	}

	fun getOsmObjectId(obj: BinaryMapDataObject?): Long {
		if (obj == null) {
			return 0
		}
		return getOsmId(obj.getId() shr 1)
	}

	fun isOsmUrlAvailable(obj: MapObject): Boolean {
		val id = obj.getId()
		return id != null && id > 0
	}

	private fun getOsmId(id: Long): Long {
		// According methods assignIdForMultipolygon and genId in IndexPoiCreator
		val clearBits = RELATION_BIT or SPLIT_BIT
		val cleaned = if (isShiftedID(id)) (id and clearBits.inv()) shr DUPLICATE_SPLIT else id
		return cleaned shr SHIFT_ID
	}

	private fun isShiftedID(id: Long): Boolean = isIdFromRelation(id) || isIdFromSplit(id)

	private fun isIdFromRelation(id: Long): Boolean = id > 0 && (id and RELATION_BIT) == RELATION_BIT

	private fun isIdFromPropagatedNode(id: Long): Boolean =
		id > 0 && (id and PROPAGATE_NODE_BIT) == PROPAGATE_NODE_BIT

	private fun isIdFromSplit(id: Long): Boolean = id > 0 && (id and SPLIT_BIT) == SPLIT_BIT

	fun isTagNonIndexedForSearchAsName(tag: String): Boolean = tag == "ref"

	fun isTagIndexedForSearchAsName(tag: String?): Boolean {
		if (tag != null) {
			// search related but not direct
			if (tag.startsWith(Amenity.ROUTE_NAME) || tag == Amenity.SHIELD_STUB_NAME) {
				return false
			}
			// some popular tags ignored as name
			if (tag.startsWith("tiger:") || tag.startsWith("noname") ||
				tag.startsWith("name:" + MapObject.NAME_ETYMOLOGY_ATTR) ||
				tag.startsWith("artist_name") ||
				tag.startsWith("addr:") // housename, street name
			) {
				return false
			}
			return tag.contains("name") || tag.contains("brand")
		}
		return false
	}

	fun isTagIndexedForSearchAsId(tag: String?): Boolean {
		if (tag != null) {
			return tag == Amenity.WIKIDATA || tag == Amenity.ROUTE_ID
		}
		return false
	}

	fun isTagIndexedAsSearchRelated(tag: String?): Boolean {
		if (tag != null) {
			return tag == Amenity.ROUTE_MEMBERS_IDS || tag == Amenity.ROUTE_NAME
		}
		return false
	}
}
